tests/js/test_utils.js

```diff
@@ -0,0 +1,238 @@
+        ),
+        [
+            {lane: '1', x: 1, run_id: 'run1', 'd2': {y: 1, run_id: 'run1'}, 'd3': {z: 'a', run_id: 'run1'}},
+            {lane: '2', x: 2, run_id: 'run1', 'd2': {y: 1, run_id: 'run1'}, 'd3': {z: 'a', run_id: 'run1'}},
+            {lane: '1', x: 3, run_id: 'run2', 'd2': {y: 4, run_id: 'run2'}, 'd3': {z: 'b', run_id: 'run2'}}
+        ]
+    );
+
+    // Test with list of keys including on null
+    assert.deepEqual(
+        merge_on_key_keep_first_sub_properties(
+            [datasrc1, datasrc2, datasrc3],
+            'run_id',
+            [null, 'd3']
+        ),
+        [
+            {lane: '1', x: 1, run_id: 'run1', y: 1, 'd3': {z: 'a', run_id: 'run1'}},
+            {lane: '2', x: 2, run_id: 'run1', y: 1, 'd3': {z: 'a', run_id: 'run1'}},
+            {lane: '1', x: 3, run_id: 'run2', y: 4, 'd3': {z: 'b', run_id: 'run2'}}
+        ]
+    );
+});
+
+
+QUnit.test('merge_multi_sources', function(assert) {
+    // patching ajax
+    var original_ajax = $.ajax;
+    var fake_ajax = function(config) {
+        var data;
+        if (config.url == 'an_api_url') {
+            data = [
+                {sample_id: 'sample_1', x: 1},
+                {sample_id: 'sample_2', x: 2}
+            ]
+        } else if (config.url == 'another_api_url') {
+            data = [
+                {sample_id: 'sample_2', y: 3},
+                {sample_id: 'sample_1', y: 4}
+            ]
+        }
+        return [{data: data}];
+    };
+    $.ajax = fake_ajax;
+    // patching complete
+
+    var observed_data;
+    var callback = function(x) {
+        observed_data = x.data;  // capture the results of merge_multi_sources
+    }
+
+    merge_multi_sources(['an_api_url', 'another_api_url'], 'token', 'sample_id')('some_data', callback, 'some_settings')
+
+    assert.deepEqual(
+        observed_data,
+        [
+            {sample_id: 'sample_1', x: 1, y: 4},
+            {sample_id: 'sample_2', x: 2, y: 3}
+        ]
+    );
+
+    $.ajax = original_ajax;  // end patch
+});
+
+QUnit.test('test_exist', function(assert) {
+    var t;
+    assert.notOk(test_exist(t));
+    assert.notOk(test_exist(null));
+    assert.notOk(test_exist(''));
+    assert.notOk(test_exist([]));
+    assert.notOk(test_exist([null]));
+    assert.ok(test_exist('1, 2, 3'));
+    assert.ok(test_exist(['1', '2', '3']));
+});
+
+//Commented out because it does not pass in phantomJS
+//QUnit.test('test_significant_figures', function(assert) {
+//    var list_number = [1, 2, 4, 7, 2, 5];
+//    assert.equal(significant_figures(list_number), 0)
+//
+//    list_number = [1.868, 0.293, 2.1274, 1.4728, 1.3457, 0.71895];
+//    assert.equal(significant_figures(list_number), 2)
+//
+//    list_number = [10.868, 67.293, 54.1274, 23.4728, 7.3457, 61.71895];
+//    assert.equal(significant_figures(list_number), 1)
+//
+//    list_number = [1245.868, 7347.293, 2684.1274, 930.4728, 2535.3457, 1651.71895];
+//    assert.equal(significant_figures(list_number), 0)
+//
+//});
```

Comment:
This is very annoying, but there's no way around it without changing how QUnit is run on the command line. If uncommented, these tests pass in web browser